docs/admin-guide/configuration.md
# Configuration

The following environment variables are available to configure the application:

| Variable                                                                          | Description                                                              | Required | Default            |
| --------------------------------------------------------------------------------- | ------------------------------------------------------------------------ | -------- | ------------------ |
| [`DISCORD_TOKEN`](#discord_token)                                                 | The Discord bot token.                                                   | Yes      | -                  |
| [`BOT_CONFIG_PATH`](#bot_config_path)                                             | The path to the bot's configuration file.                                | No       | `application.yaml` |
| [`DISCORD_API_CONCURRENCY`](#discord_api_concurrency)                             | The maximum number of concurrent Discord API requests.                   | No       | `3`                |
| [`ENVIRONMENT`](#environment)                                                     | The environment in which the application is running.                     | No       | `production`       |
| [`HF_DOWNLOAD_CONCURRENCY`](#hf_download_concurrency)                             | The maximum number of concurrent downloads when installing transformers. | No       | `3`                |
| [`HF_HOME`](#hf_home)                                                             | The directory containing Huggingface Transformers data files.            | No       | `hf_data`          |
| [`LOG_LEVEL`](#log_level)                                                         | The minimum log level.                                                   | No       | `INFO`             |
| [`MPLCONFIGDIR`](#mplconfigdir)                                                   | The directory containing Matplotlib configuration files.                 | No       | `/app/matplotlib`  |
| [`NLTK_DATA`](#nltk_data)                                                         | The directory containing NLTK data files.                                | No       | `nltk_data`        |
| [`NLTK_DOWNLOAD_CONCURRENCY`](#nltk_download_concurrency)                         | The maximum number of concurrent downloads when installing NLTK data.    | No       | `3`                |
| [`PREPROCESSING_MAX_WORD_LENGTH`](#preprocessing_max_word_length)                 | The maximum word length. Longer words are dropped.                       | No       | `35`               |
| [`PREPROCESSING_MESSAGE_TRUNCATE_LENGTH`](#preprocessing_message_truncate_length) | The maximum message length. Longer messages are truncated.               | No       | `256`              |
| [`REDIS_HOST`](#redis_host)                                                       | The Redis host.                                                          | No       | `localhost`        |
| [`REDIS_PORT`](#redis_port)                                                       | The Redis port.                                                          | No       | `6379`             |
| [`REDIS_PASSWORD`](#redis_password)                                               | The Redis password.                                                      | No       | -                  |

## Required Settings

The following settings are required to start the application:

### `DISCORD_TOKEN`

You can obtain a Discord bot token from the [Discord Developer Portal](https://discord.com/developers/applications).
Your token should have the following scopes:

- `bot`

!!! DANGER "Security Warning"

    Do not share your token with anyone!

## Optional Settings

The following settings are optional or have default values that can be overridden:

### `BOT_CONFIG_PATH`

This specifies the location of the bot's configuration file. By default, the application searches for a file named
`application.yaml` in the directory from which it is launched. In the Docker image, this file is located at `/app/application.yaml`.

[Read more](#applicationyaml) about the `application.yaml` file.

### `DISCORD_API_CONCURRENCY`

The maximum number of concurrent Discord API requests. By default, this is set to `3`.

### `ENVIRONMENT`

The environment in which the application is running. Set this to `development` to enable development features
such as loading development cogs. By default, this is set to `production`.

### `HF_DOWNLOAD_CONCURRENCY`

The maximum number of concurrent downloads when installing Huggingface Transformers models. By default, this
is set to `3`.

### `HF_HOME`

The directory containing Huggingface Transformers data files. By default, this is set to `hf_data` in the directory
from which the application is launched. In the Docker image, this directory is located at `/app/hf_data`.

### `LOG_LEVEL`

The minimum log level to display. The following levels are available:

- `DEBUG`
- `INFO`
- `WARNING`
- `ERROR`
- `CRITICAL`

The default log level is `INFO`.

### `MPLCONFIGDIR`

The directory containing `matplotlib` configuration files. Uses the default `matplotlib` configuration directory
unless otherwise configured. In the Docker image, the `matplotlib` configuration files are located at `/app/matplotlib`.

### `NLTK_DATA`

The directory containing NLTK data files. By default, this is set to `nltk_data` in the directory from which the
application is launched. In the Docker image, this directory is located at `/app/nltk_data`.

### `NLTK_DOWNLOAD_CONCURRENCY`

The application automatically downloads NLTK data files on startup. This setting controls the number of concurrent
downloads. By default, this is set to `3`.

### `PREPROCESSING_MAX_WORD_LENGTH`

The maximum word length. Words longer than this value are dropped. By default, this is set to `35`.

### `PREPROCESSING_MESSAGE_TRUNCATE_LENGTH`

The maximum message length. Messages longer than this value are truncated. By default, this is set to `256`.

### `REDIS_HOST`

The hostname of the Redis server. Defaults to `localhost`.

### `REDIS_PORT`

The port of the Redis server. Defaults to `6379`.

### `REDIS_PASSWORD`

The password of the Redis server. Set this variable if your Redis server requires authentication. No password
is set by default.

!!! DANGER "Security Warning"

    Do not share your Redis password with anyone!

## `application.yaml`

The `application.yaml` file is a configuration file that specifies the cogs to load, the NLTK datasets to download,
and the Huggingface Transformers models to install on startup. It has the following structure:

```yaml
# List of cogs to load when the bot starts, identified by their package name.
cogs:
  - <PACKAGE_NAME>
  - <PACKAGE_NAME>
# List of cogs to load in development mode only, identified by their package name.
dev-cogs:
  - <PACKAGE_NAME>
  - <PACKAGE_NAME>
# List of NLTK datasets to download on startup.
nltk:
  - <DATASET_NAME>
  - <DATASET_NAME>
# List of Huggingface Transformers models to download on startup.
transformers:
  - <MODEL_NAME>
  - <MODEL_NAME>
```

!!! WARNING "Advanced Users Only"

    For standard use, the `application.yaml` file does not need to be changed. We recommend against modifying this
    file unless you know what you're doing.

One possible use case is to add additional feature or disable existing features by modifying the list of cogs.
To add a new cog, add the package name to the `cogs` list. To disable a feature, remove the package name from
the list.

Below is a list of all cogs that are loaded by default:

| Package Name                                            | Description                                                                      |
| ------------------------------------------------------- | -------------------------------------------------------------------------------- |
| `courageous_comets.cogs.about`                          | Provides information about the bot.                                              |
| `courageous_comets.cogs.keywords.search_command`        | Searches for keywords using a slash command.                                     |
| `courageous_comets.cogs.keywords.search_context_menu`   | Searches for keywords using a context menu item.                                 |
| `courageous_comets.cogs.keywords.topics_command`        | Lists the most popular keywords for a given context using a slash command.       |
| `courageous_comets.cogs.keywords.user_context_menu`     | Lists the most popular keywords for a particular user using a context menu item. |
| `courageous_comets.cogs.messages`                       | Listens for messages and passes them on for internal processing.                 |
| `courageous_comets.cogs.ping`                           | Responds to ping requests.                                                       |
| `courageous_comets.cogs.sentiment.message_context_menu` | Analyzes the sentiment of a message using a context menu item.                   |
| `courageous_comets.cogs.sentiment.search_command`       | Searches for messages with similar sentiment using a slash command.              |
| `courageous_comets.cogs.sentiment.search_context_menu`  | Searches for messages with similar sentiment using a context menu item.          |
| `courageous_comets.cogs.sentiment.user_context_menu`    | Analyzes the sentiment of a user's messages using a context menu item.           |
| `courageous_comets.cogs.frequency`                      | Provides an overview of the amount of recent messages for a given timespan.      |


docs/contributor-guide/development-environment.md
# Development Environment

Follow the steps below to set up your development environment.

!!! NOTE "Prerequisites"

    You need to have [Git](https://git-scm.com) installed on your system.

## Environment Setup

You can set up the development environment using either the [development container](#using-the-development-container)
or following the [manual](#manual-setup) setup process.

### Using the Development Container

The project includes a [development container](https://containers.dev) to automatically set up your development
environment, including the all tools and dependencies required to develop the application locally.

!!! NOTE "Prerequisites"

    [Docker](https://www.docker.com) must be installed on your system to use the development container.

#### Quick Start

See the video installation guide below for a step-by-step tutorial on installing the development container with
Visual Studio Code:

<video controls>
    <source src="https://github.com/user-attachments/assets/703aa245-9e33-44d9-9c79-7432afbeb445" type="video/mp4">
</video>

First, install the [Remote Development Extension Pack](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack).
Next, open Visual Studio Code and click on icon in the bottom left corner to open the command palette for remote
environments.

Select the `Clone Repository in Container Volume` command. This will prompt you to select the
repository to clone. Choose the `thijsfranck/courageous-comets` repository or paste the repository URL.

Once you confirm the selection, the development container will be set up automatically.

#### Detailed Setup Guide

For more details, refer to the setup guide for your IDE:

- [Visual Studio Code](https://code.visualstudio.com/docs/devcontainers/tutorial)
- [PyCharm](https://www.jetbrains.com/help/pycharm/connect-to-devcontainer.html)

#### Services

The development container includes the following services for local development:

| Service      | Description  | Address                                   |
| ------------ | ------------ | ----------------------------------------- |
| Redis        | Database     | [`localhost:6379`](http://localhost:6379) |
| RedisInsight | Database GUI | [`localhost:8001`](http://localhost:8001) |

### Manual Setup

If you prefer to set up the development environment manually, follow the steps below.

!!! NOTE "Prerequisites"

    Please ensure [Python 3.12](https://www.python.org) and [Poetry](https://python-poetry.org) are installed
    on your system.

#### Clone the Repository

To clone the repository, run the following command:

```bash
git clone https://github.com/thijsfranck/courageous-comets.git
```

Next, open the project in your preferred IDE or navigate to the project directory using the terminal.

#### Install Dependencies

Start by installing the project dependencies using Poetry:

```bash
poetry install
```

This will create a virtual environment and install the required dependencies.

#### Pre-commit Hooks

Next, install the pre-commit hooks to ensure that your code is formatted and linted before each commit:

```bash
poetry run pre-commit install
```

This will set up the pre-commit hooks to run automatically when you commit changes to the repository.

#### Redis Database

The application requires a Redis database to run. We recommend setting up a local Redis instance using Docker
for development purposes.

To start a Redis instance using Docker, run the following command:

```bash
docker run -d -p 6379:6379 -p 8001:8001 redis/redis-stack:latest
```

This will start a Redis server on port `6379` and a RedisInsight GUI on port `8001`.

## Configuring your Environment

To run the application, you will need to provide the following configurations in a `.env` file at the project
root directory.

### Discord Token

The application requires a Discord bot token to run. It should be stored in the `.env` file as follows:

```dotenv
DISCORD_TOKEN=<YOUR DISCORD_TOKEN>
```

The repository includes an encrypted `.env.lock` file with the shared Discord bot token for our team. Follow the
[Secrets Management](./secrets-management.md) guide to decrypt the file and start using the token.

??? QUESTION "Can I use my own Discord bot token?"

    Yes, you can use your own Discord bot token. If you do so, there's need to decrypt the `.env.lock` file.

### Redis Configuration

If you're not using the development container, you will need to configure the Redis connection in the `.env` file:

```dotenv
REDIS_HOST=localhost
REDIS_PORT=6379
```

This configuration assumes you are running a local Redis instance on the default port.

## Running the Application

With your development environment set up and configured, you can run the application using the following command:

```bash
poetry run python -m courageous_comets
```

The application should now be online and ready to respond to input from your Discord server.

## Building the Docker Image

!!! INFO "Production Builds"

    The release process is fully automated and does not require you to build the docker image locally. See the
    [GitHub Actions](./version-control.md#github-actions) section of the version control guide for more information.

Before building the Docker image, first build the Python package with Poetry:

```bash
poetry build -f wheel
```

This will create a `.whl` file in the `dist` directory. Next, build the Docker image as follows:

```bash
docker build -t ghcr.io/thijsfranck/courageous-comets:latest .
```

## Running the Docker Container

Once you have [built the Docker image](#building-the-docker-image), use the following command to run the production
container locally:

```bash
docker run -i --env-file .env ghcr.io/thijsfranck/courageous-comets:latest
```

This will run the application just as it would in production, using your local `.env` file and Redis instance.

## Running the Docker Compose Stack

To run the application in a production configuration including the Redis database, you can use the Docker Compose
stack.

First, build the Docker image as described in [the previous section](#building-the-docker-image). Then, run the
following command:

```bash
docker-compose up
```

This will start the application and the Redis database in separate containers using your local `.env` file.

## Running the Documentation

To view the documentation locally, you can use the following command:

```bash
poetry run mkdocs serve
```

Open your browser and navigate to [`http://localhost:8000`](http://localhost:8000) to view the documentation.
The changes you make to the documentation will be automatically reflected in the browser.


docs/CHANGELOG.md
## v1.0.0 (2024-07-28)

### BREAKING CHANGE

- dependencies are no longer updated

### Feat

- lock dependencies for final version

## v0.10.0 (2024-07-28)

### Feat

- frequency time divisions (#62)
- add topics command that lists the most used keywords (#61)
- add user keyword frequency (#60)

### Fix

- capitalize embed title
- ensure consistent styling on embed titles
- improve logging for all interactions

### Refactor

- log message processed as debug event
- log database event as error
- simplify messages cog
- remove redundant error class
- remove redundant decorator
- split up keywords and sentiment cogs, improve cog documentation

## v0.9.0 (2024-07-28)

### Feat

- add praise interaction for user and message sentiment
- harmonize sentiment embeds
- add details to frequency embed
- refactor cogs with ui components and complete existing interactions

### Fix

- increase search results max preview length
- show channel in search results
- avoid error when no sentiment data is available for a user

### Refactor

- remove redundant code
- remove unused component
- move message processing to a separate module

## v0.8.0 (2024-07-27)

### Feat

- plot message frequency graph

### Fix

- reference current version in documentation
- update app description and set embed footer
- handle missing sentiment data
- restore user sentiment interaction
- add exists check
- silence KeyboardInterrupt on application shutdown (#59)
- remove  cog from erroneous merge conflict resolution (#58)
- remove unused context menu loader

## v0.7.0 (2024-07-26)

### Feat

- add user sentiment interaction (#54)
- set up frequency cog (#53)
- add sentiment chart (#52)
- add cog that returns information about the app (#50)
- add drop_code_blocks processor
- get message rate by duration (#49)
- calculate count of tokens across messages (#48)
- allow redis search filtering across multiple ids (#47)
- add development mode (#46)

### Fix

- remove reference to link from general bot description and add it to the about message
- handle PackageNotFoundError
- add __version__ and log version on startup

## v0.6.0 (2024-07-24)

### Feat

- store message tokens (#45)
- limit queries to most recent messsages (#43)
- store sentiment results along with message (#42)

### Refactor

- simplify function signatures and make it easier to build scope filters (#44)

## v0.5.0 (2024-07-23)

### Feat

- add docker compose and fix production dockerfile

## v0.4.0 (2024-07-23)

### Feat

- revert back to hash data type for storing messages (#41)

### Fix

- open app config in read mode
- trucate uses correct length
- preprocessing drops extra whitespace

## v0.3.0 (2024-07-22)

### Feat

- store sentiment analysis with message data (#39)
- set up message content preprocessing (#40)
- replace sentence_transformers with transformers library (#36)
- add messages cog (#32)
- return messages based on similarity score (#30)

### Fix

- fix sync command signature and internal logic (#38)
- save messages with message_id on redis (#33)
- ensure consistent log output (#29)
- suppress warnings from libraries

### Refactor

- remove hfvectorizer (#35)
- include save_message in Messages cog (#34)
- remove throws clause from docstring

## v0.2.0 (2024-07-21)

### Feat

- setup transformer models (#28)
- use pydantic to model Redis hashes (#25)
- add sentiment analysis (#24)
- add api to store word frequency on redis (#23)
- download nltk resources on startup (#22)

### Fix

- handle contractions in tokenizer (#26)
- control max number of concurrent downloads
- add logging and remove return section from docstring

### Refactor

- move startup logic into bot class (#27)
- improve docs, types and fix some linter issues
- update docstrings
- improve type hints

## v0.1.0 (2024-07-20)

### Feat

- add bot boilerplate

### Fix

- improved error handling and settings validation
- include application config in docker image (#15)
- set redis host from environment variables

### Refactor

- improve error handling and logging

## v0.0.6 (2024-07-19)

## v0.0.5 (2024-07-17)

## v0.0.4 (2024-07-16)

## v0.0.3 (2024-07-16)

## v0.0.2 (2024-07-16)

## v0.0.1 (2024-07-15)


courageous_comets/settings.py
"""Shared application configuration."""

import logging
import os
import sys
import warnings
from pathlib import Path

import coloredlogs
from dotenv import load_dotenv

from courageous_comets.exceptions import ConfigurationValueError


def read_bot_config_path() -> Path:
    """
    Read the bot configuration path from the environment.

    Returns
    -------
    pathlib.Path
        The path to the bot configuration file.

    Raises
    ------
    courageous_comets.exceptions.ConfigurationValueError
        If the path does not exist.
    """
    result = Path(os.getenv("BOT_CONFIG_PATH", "application.yaml"))

    if not result.exists():
        raise ConfigurationValueError(
            key="BOT_CONFIG_PATH",
            value=os.getenv("BOT_CONFIG_PATH"),
            reason=f"No file found at path '{result.resolve()}'",
        )

    return result


def read_discord_token() -> str:
    """
    Read the Discord token from the environment.

    Returns
    -------
    str
        The Discord bot token.

    Raises
    ------
    courageous_comets.exceptions.ConfigurationValueError
        If the token is not found
    """
    result = os.getenv("DISCORD_TOKEN", "")

    if not result:
        raise ConfigurationValueError(
            key="DISCORD_TOKEN",
            value=os.getenv("DISCORD_TOKEN"),
            reason="No Discord bot token found. Set the DISCORD_TOKEN environment variable.",
        )

    return result


def read_int(key: str, default: int) -> int:
    """
    Read an integer value from the environment.

    Parameters
    ----------
    key : str
        The environment variable key.
    default : int
        The default value to use if the environment variable is not set.

    Returns
    -------
    int
        The integer value.

    Raises
    ------
    courageous_comets.exceptions.ConfigurationValueError
        If the value is not a valid integer.
    """
    try:
        result = int(os.getenv(key, str(default)))
    except ValueError as e:
        raise ConfigurationValueError(
            key=key,
            value=os.getenv(key),
            reason="Value must be an integer",
        ) from e

    return result


def read_redis_port() -> int:
    """
    Read the Redis port from the environment.

    Returns
    -------
    int
        The Redis port number.

    Raises
    ------
    courageous_comets.exceptions.ConfigurationValueError
        If the port is not a valid port number.
    """
    result = read_int("REDIS_PORT", 6379)

    if not (0 <= result <= 65535):  # noqa: PLR2004
        raise ConfigurationValueError(
            key="REDIS_PORT",
            value=result,
            reason="Value must be a valid port number (0-65535)",
        )

    return result


def setup_logging() -> None:
    """Set up logging for the application."""
    coloredlogs.install(
        level=LOG_LEVEL,
        fmt="{asctime} {levelname:<8} [{name}] {message}",
        datefmt="%Y-%m-%d %H:%M:%S",
        style="{",
        reconfigure=True,
    )


# Load environment variables from a .env file. If the file does not exist, this does nothing.
load_dotenv()

# Suppress warnings from libraries
warnings.filterwarnings("ignore")

# Set up logging for the application
LOG_LEVEL = logging.getLevelNamesMapping().get(
    os.getenv("LOG_LEVEL", "INFO"),
    logging.INFO,
)

setup_logging()

# Determine whether the application is running in a development environment
IS_DEV = os.getenv("ENVIRONMENT", "production") == "development"

if IS_DEV:
    logging.warning("🚨 Application is running in development mode! 🚨")

# Load configuration values from the environment
try:
    DISCORD_TOKEN = read_discord_token()
    BOT_CONFIG_PATH = read_bot_config_path()
    DISCORD_API_CONCURRENCY = read_int("DISCORD_API_CONCURRENCY", 3)
    NLTK_DATA_DIR = os.getenv("NLTK_DATA", "nltk_data")
    NLTK_DOWNLOAD_CONCURRENCY = read_int("NLTK_DOWNLOAD_CONCURRENCY", 3)
    PREPROCESSING_MAX_WORD_LENGTH = read_int("MAX_WORD_LENGTH", 35)
    PREPROCESSING_MESSAGE_TRUNCATE_LENGTH = read_int("TRUNCATE_LENGTH", 256)
    REDIS_HOST = os.getenv("REDIS_HOST", "localhost")
    REDIS_PORT = read_redis_port()
    REDIS_PASSWORD = os.getenv("REDIS_PASSWORD")
    REDIS_KEYS_PREFIX = os.getenv("REDIS_KEYS_PREFIX", "courageous_comets")
    # Maximum number of items to return from a query
    QUERY_LIMIT = read_int("QUERY_LIMIT", 10)
    # Huggingface environment variable for caching downloaded models.
    # https://huggingface.co/docs/huggingface_hub/v0.24.0/package_reference/environment_variables#hf_home
    HF_HOME = os.getenv(
        "HF_HOME",
        "hf_data",
    )
    HF_DOWNLOAD_CONCURRENCY = read_int("HF_DOWNLOAD_CONCURRENCY", 3)
except ConfigurationValueError as e:
    logging.critical(
        "Cannot start the application due to configuration errors",
        exc_info=e,
    )
    sys.exit(1)

logging.info("Configuration loaded successfully")
